lib/Mojolicious.pm

```diff
@@ -0,0 +1,1217 @@
+actions or handle return values differently. Note that this hook can trigger more than once for the same request if
+there are nested routes. (Passed a callback leading to the next hook, the current controller object, the action
+callback and a flag indicating if this action is an endpoint)
+
+=head2 before_render
+
+Emitted before content is generated by the renderer. Note that this hook can trigger out of order due to its dynamic
+nature, and with embedded applications will only work for the application that is rendering.
+
+  $app->hook(before_render => sub ($c, $args) {...});
+
+Mostly used for pre-processing arguments passed to the renderer. (Passed the current controller object and the render
+arguments)
+
+=head2 after_render
+
+Emitted after content has been generated by the renderer that will be assigned to the response. Note that this hook can
+trigger out of order due to its dynamic nature, and with embedded applications will only work for the application that
+is rendering.
+
+  $app->hook(after_render => sub ($c, $output, $format) {...});
+
+Mostly used for post-processing dynamically generated content. (Passed the current controller object, a reference to
+the content and the format)
+
+=head2 after_dispatch
+
+Emitted in reverse order after a response has been generated. Note that this hook can trigger out of order due to its
+dynamic nature, and with embedded applications will only work for the application that is generating the response.
+
+  $app->hook(after_dispatch => sub ($c) {...});
+
+Useful for rewriting outgoing responses and other post-processing tasks. (Passed the current controller object)
+
+=head1 ATTRIBUTES
+
+L<Mojolicious> implements the following attributes.
+
+=head2 commands
+
+  my $commands = $app->commands;
+  $app         = $app->commands(Mojolicious::Commands->new);
+
+Command line interface for your application, defaults to a L<Mojolicious::Commands> object.
+
+  # Add another namespace to load commands from
+  push @{$app->commands->namespaces}, 'MyApp::Command';
+
+=head2 controller_class
+
+  my $class = $app->controller_class;
+  $app      = $app->controller_class('Mojolicious::Controller');
+
+Class to be used for the default controller, defaults to L<Mojolicious::Controller>. Note that this class needs to have
+already been loaded before the first request arrives.
+
+=head2 exception_format
+
+  my $format = $app->exception_format;
+  $app       = $app->exception_format('txt');
+
+Format for HTTP exceptions (C<html>, C<json>, or C<txt>), defaults to C<html>.
+
+=head2 home
+
+  my $home = $app->home;
+  $app     = $app->home(Mojo::Home->new);
+
+The home directory of your application, defaults to a L<Mojo::Home> object which stringifies to the actual path.
+
+  # Portably generate path relative to home directory
+  my $path = $app->home->child('data', 'important.txt');
+
+=head2 log
```

Comment:
It would also be nice for this attribute to default the log level to a setting defined in the config file, and also to enable STDERR even when a path is specified -- or, if the best/only way to do that is via message events, document that perhaps here (or save it for the cookbook -- perhaps an events section).